### ISSUE TITLE
S1854 FP: `try-catch` inside a loop

### ISSUE THREAD

**Issue by remcoros** [SonarSource/sonar-dotnet#2393]
### Description

I'm getting a false positive on S1854

### Repro steps

```csharp
    public class Foo
    {
        public void Bar()
        {
            bool isFirst = true;
            foreach (var i in Enumerable.Range(1, 10))
            {
                try
                {
                    if (isFirst)
                    {
                        Console.WriteLine("First");
                    }
                    else
                    {
                        Console.WriteLine("Not first");
                    }
                }
                finally
                {
                    isFirst = false;
                }
            }
        }
    }
```

### Expected behavior

No warning

### Actual behavior

![image](https://user-images.githubusercontent.com/130373/56343966-8be07600-61bc-11e9-9f92-ebe477a37ad4.png)

### Known workarounds

n/a

### Related information

* SonarLint for VS Version: 4.9.0.4316
* Visual Studio 2019 / C#

**Comment by Corniel** [comment]
To me this is valid warning. The finally will only be reached to clean up some stuff. Afterwards, the Exception is still thrown.

**Comment by remcoros** [comment]
It's a simple example (I didn't add a 'catch' in this), but it makes no difference if I add it.

If I remove this "useless assignment", the loop/isFirst check doesn't work anymore, so it's not a valid suggestion.

**Comment by Corniel** [comment]
Ah, now I see. it is within a `foreach` loop. Me bad. So yes, it is a false positive. However, in real examples, I would argue it would be better to put the complex try-catch code within the loop in an separate function to improve readability.

**Comment by remcoros** [comment]
Yes, you're absolutely right. I just wanted to capture the issue here in a minimal example.

**Comment by andrei-epure-sonarsource** [comment]
Adding an example with try-catch FPs inside a loop from @killergege as mentioned in #697

```
public static void RetryOnException(int retries, TimeSpan delay, Action operation)
{
    var attempts = 0;
    do
    {
        try
        {
            attempts++; //Remove this useless assignment to local variable 'attempts'.
            operation();
            break;
        }
        catch (Exception ex)
        {
            if (attempts > retries)
                throw;
            Logger.LogError($"Exception caught on attempt {attempts} - will retry after delay {delay}", ex);
            Task.Delay(delay).Wait();
        }
    } while (true);
}
```
